tasmota/berry/drivers/Core200S.be

```diff
@@ -0,0 +1,712 @@
+                if self.value_map[k] == value
+                    self.value = value
+                    break
+                end
+            end
+        elif type(value) == 'string'
+            if self.value_map.contains(value)
+                self.value = self.value_map[value]
+            end
+        end
+
+        if self.value == nil
+            raise "StateEnumError", "Invalid type or value"
+        end
+    end
+
+    def tostring()
+        for k : self.value_map.keys()
+            if self.value_map[k] == self.value
+                return k
+            end
+        end
+        return "invalid"
+    end
+
+    def toint()
+        return self.value
+    end
+
+    def ==(other)
+        return self.value == other.value
+    end
+
+end
+
+class OnOffState : StateEnum
+    static ON = OnOffState("on")
+    static OFF = OnOffState("off")
+    def init(value)
+        super(self).init({"off": 0, "on": 1}, value)
+    end
+end
+
+class FanMode : OnOffState
+    def init(value)
+        super(self).init(value)
+    end
+end
+
+class ChildLockMode : OnOffState
+    def init(value)
+        super(self).init(value)
+    end
+end
+
+class SleepMode : OnOffState
+    def init(value)
+        super(self).init(value)
+    end
+end
+
+class DisplayAutoOffMode : StateEnum
+    static OFF = DisplayAutoOffMode("off")
+    static ON = DisplayAutoOffMode("on")
+    def init(value)
+        super(self).init({"off": 1, "on": 0}, value)
+    end
+end
+
+class DisplayMode : StateEnum
+    def init(value)
+        super(self).init({"off": 0, "on": 100}, value)
+    end
+end
+
+class FanSpeedMode : StateEnum
+    static LOW = FanSpeedMode("low")
+    static MED = FanSpeedMode("med")
+    static FULL = FanSpeedMode("full")
+
+    def init(value)
+        super(self).init({"low": 1, "med": 2, "full": 3}, value)
+    end
+end
+
+class NightLightMode : StateEnum
+    static OFF = NightLightMode("off")
+    static LOW = NightLightMode("low")
+    static FULL = NightLightMode("full")
+    def init(value)
+        super(self).init({"off": 0, "low": 50, "full": 100}, value)
+    end
+end
+
+def modulo(num, mod)
+    return ((num % mod) + mod) % mod
+end
+
+def info_c2s(msg)
+    log("C2S: " + msg, 2)
+end
+
+def log_c2s(msg)
+    log("C2S: " + msg)
+end
+
+class Core200SSerial : serial
+    var msg_count
+    var read_buffer
+    var error_count
+    var expected_replies
+
+    def init()
+        super(self).init(16, 17, 115200)
+        self.reset_counters()
+        self.read_buffer = bytes("")
+    end
+
+    def deinit()
+        self.close()
+    end
+
+    static def calc_checksum(msg)
+        var checksum = 0xff
+        var i = 0
+        while i < msg.size()
+            if (i != 5)
+                checksum -= msg.get(i)
+                checksum = modulo(checksum, 256)
+            end
+            i += 1
+        end
+        return checksum
+    end
+
+    def send(msg)
+        self.send_nowait(msg)
+        tasmota.delay(100)
```

Comment:
This is not the right way to do it. `tasmota.delay(100)` stops the entire Tasmota for 100ms which is much higher than the standard loop.

Prefer `tasmota.set_timer(100, <your_function>)` for asynchronous callback

Reply:
hmm I'm not entirely sure how this is supposed to work. `send()` will eventually be called by all of the commands. If we switch to asynchronous handling here, how am I supposed to report if the command was successful or not?

Reply:
Generally speaking, you first acknowledge the command, and then you post a message once the message is accepted.